### PR TITLE
New YAML parser interface

### PR DESCRIPTION
**Pull request by dkeeney**
Continuation of PR #282

### CODE REVIEW

File: src/htm/ntypes/Value.hpp [hunk 1]
```diff
@@ -14,122 +16,236 @@
...
+#include <htm/utils/Log.hpp>
 #include <string>
+#include <vector>
+#include <map>
+#include <memory>
+#include <iostream>
+#include <sstream>
+#include <algorithm>  // transform
+#include <cerrno>
+#include <cstring>    // std::strerror(errno)
```
Comment (breznak) [inline src/htm/ntypes/Value.hpp]: nit, this seems to define a lot of headers, is all of that really needed, or can some be moved to cpp?

File: src/htm/regions/SPRegion.cpp [hunk 2]
```diff
@@ -46,7 +45,7 @@ SPRegion::SPRegion(const ValueMap &values, Region *region)
   args_.columnCount = values.getScalarT<UInt32>("columnCount", 0);
   args_.potentialRadius = values.getScalarT<UInt32>("potentialRadius", 0);
   args_.potentialPct = values.getScalarT<Real32>("potentialPct", 0.5);
-  args_.globalInhibition = values.getScalarT<bool>("globalInhibition", true);
+  args_.globalInhibition = values.getScalarT<bool>("globalInhibition", false);
```
Comment (breznak) [inline src/htm/regions/SPRegion.cpp]: please use True as default, local inh is quite slow, and global is the default in the SP class.

File: src/htm/regions/SPRegion.cpp [hunk 3]
```diff
@@ -55,7 +54,7 @@ SPRegion::SPRegion(const ValueMap &values, Region *region)
   args_.minPctOverlapDutyCycles = values.getScalarT<Real32>("minPctOverlapDutyCycles", 0.001f);
   args_.dutyCyclePeriod = values.getScalarT<UInt32>("dutyCyclePeriod", 1000);
   args_.boostStrength = values.getScalarT<Real32>("boostStrength", 0.0f);
-  args_.seed = values.getScalarT<Int32>("seed", 1);
+  args_.seed = values.getScalarT<Int32>("seed", -1);
```
Comment (breznak) [inline src/htm/regions/SPRegion.cpp]: why -1? Rather keep no-change, as this shouldn't affect this PR in any way.

File: src/htm/regions/ScalarSensor.cpp [hunk 4]
```diff
@@ -32,19 +32,19 @@ namespace htm {
...
-  params_.activeBits = params.getScalarT<UInt32>("w");
-  params_.resolution = params.getScalarT<Real64>("resolution");
-  params_.radius = params.getScalarT<Real64>("radius");
-  params_.minimum = params.getScalarT<Real64>("minValue");
-  params_.maximum = params.getScalarT<Real64>("maxValue");
-  params_.periodic = params.getScalarT<bool>("periodic");
-  params_.clipInput = params.getScalarT<bool>("clipInput");
+  params_.size = params.getScalarT<UInt32>("n", 0);
+  params_.activeBits = params.getScalarT<UInt32>("w", 0);
+  params_.resolution = params.getScalarT<Real64>("resolution", 0.0);
```
Comment (breznak) [inline src/htm/regions/ScalarSensor.cpp]: these 3 values of "0" are invalid combination, is it what is expected?

File: src/htm/regions/TMRegion.cpp [hunk 5]
```diff
@@ -47,7 +47,7 @@ TMRegion::TMRegion(const ValueMap &params, Region *region)
   args_.activationThreshold = params.getScalarT<UInt32>("activationThreshold", 13);
   args_.initialPermanence = params.getScalarT<Real32>("initialPermanence", 0.21f);
   args_.connectedPermanence = params.getScalarT<Real32>("connectedPermanence", 0.50f);
-  args_.minThreshold = params.getScalarT<UInt32>("minThreshold", 10);
+  args_.minThreshold = params.getScalarT<UInt32>("minThreshold", 8);
```
Comment (breznak) [inline src/htm/regions/TMRegion.cpp]: ditto, please make this correspond to what we have in TM class

File: src/test/unit/engine/CppRegionTest.cpp [hunk 6]
```diff
@@ -168,7 +168,6 @@ TEST(CppRegionTest, testYAML) {
 
   Network net;
   std::shared_ptr<Region> level1;
-  EXPECT_THROW(net.addRegion("level1", "TestNode", badparams), exception);
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: why is this removed?

File: src/htm/ntypes/Value.hpp (hunk 1 above)
Comment (dkeeney) [reply src/htm/ntypes/Value.hpp]: I am sure I can remove some when I clean up.

File: src/htm/regions/SPRegion.cpp (hunk 2 above)
Comment (dkeeney) [reply src/htm/regions/SPRegion.cpp]: >>please use True as default,

Well...the problem is that this PR does change behavior slightly.  In the past the defaults from the Spec were incorporated in `values`.  The default given here was never used.  The new behavior is that the defaults in the Spec are NOT included in `values` and therefore this default is now being used.

What I did was change these values to be the same as what was given in the Spec so that the default parameters passed on to the SP algorithm are unchanged.

Now, if you want to change the default to True in this case, then you must also change the expected results in the tests.   I agree that the default here should be the same as the defaults used by SP but that sounds like a separate PR.

File: src/htm/regions/SPRegion.cpp (hunk 3 above)
Comment (dkeeney) [reply src/htm/regions/SPRegion.cpp]: >> why -1? Rather keep no-change, 

Again, the default actually used was -1 from the Spec.  I changed it here to match so that the test results would not be affected now that the default is now taken from here.

File: src/htm/regions/ScalarSensor.cpp (hunk 4 above)
Comment (dkeeney) [reply src/htm/regions/ScalarSensor.cpp]: > these 3 values of "0" are invalid combination, is it what is expected?

Again, taking the defaults from the Spec.  In this case the algorithm also defaults to all 0's.

File: src/htm/regions/TMRegion.cpp (hunk 5 above)
Comment (dkeeney) [reply src/htm/regions/TMRegion.cpp]: >> ditto, please make this correspond to what we have in TM class

ditto to why I am keeping them the same as in the Spec.
If I do not, the unit tests fail :)

File: src/test/unit/engine/CppRegionTest.cpp (hunk 6 above)
Comment (dkeeney) [reply src/test/unit/engine/CppRegionTest.cpp]: > why is this removed?      (talking about the check for bad parameters)

The new behavior is that we do not use the Spec to evaluate the parameters so we have no way to know if 'unexpected' parameters are presented.

It is possible to restore the original behavior....on a per Region basis.  In the base class for all Region implementations `RegionImpl.hpp`  there is a new function that can be called to pre-process the parameters.
```
  virtual ValueMap ValidateParameters(const ValueMap &vm, Spec* ns);
```
This will read the Spec, insert the default values into the ValueMap, and validate for unexpected parameters. Each implementation has the option to call this in its constructor before it handles the parameters.  In an attempt to simplify everything I am first trying to do parameter processing without using this function but it is there as a fallback.

File: src/htm/regions/SPRegion.cpp (hunk 2 above)
Comment (breznak) [reply src/htm/regions/SPRegion.cpp]: > What I did was change these values to be the same as what was given in the Spec so that the default parameters passed on to the SP algorithm are unchanged. 

ok, thanks for explanation, this is fine. also for other on the same topic :+1: 

> I agree that the default here should be the same as the defaults used by SP but that sounds like a separate PR. 

yep, let's keep that for another pr

File: src/htm/regions/ScalarSensor.cpp (hunk 4 above)
Comment (breznak) [reply src/htm/regions/ScalarSensor.cpp]: > In this case the algorithm also defaults to all 0's. 

strange, I think all-0 should throw an assert...

File: src/test/unit/engine/CppRegionTest.cpp (hunk 6 above)
Comment (breznak) [reply src/test/unit/engine/CppRegionTest.cpp]: > In an attempt to simplify everything I am first trying to do parameter processing without using this function but it is there as a fallback.
 
i think it's fine not to check that, so do as you do now. 

> so we have no way to know if 'unexpected' parameters are presented.

just what happens if I pass a param `canIBreakStuff = True` to a SPRegion?

File: src/htm/regions/ScalarSensor.cpp (hunk 4 above)
Comment (dkeeney) [reply src/htm/regions/ScalarSensor.cpp]: Well, look at it in another way... You want to insure that the user **Always** specifies at least one of the combinations. If the algorithm still gets all 0's then it should throw an exception.  So taking all defaults is not allowed.

File: src/test/unit/engine/CppRegionTest.cpp (hunk 6 above)
Comment (dkeeney) [reply src/test/unit/engine/CppRegionTest.cpp]: > just what happens if I pass a param canIBreakStuff = True to a SPRegion?

It will be ignored.

**Comment by dkeeney** [comment]
@breznak I need your help.  I know I must be doing something stupid but I cannot find it.

When I run the python unit tests I get 4 errors in NetworkTest.py.  lines 164, 244, 264, 328.  All three are the same error.
```
>     r = net.addRegion("test", "py.LinkRegion", "")
E     RuntimeError: TypeError: __init__() got multiple values for argument 'self'

J:\Projects\AI\htmG\bindings\py\tests\regions\network_test.py:328: RuntimeError
```
I need someone with more Python experience to take a look at this.  It could be something to do with the bindings for addRegion( ) but I made no changes to it.

Where I did make changes is in `bindings\py\cpp_src\plugin\PyBindRegion.cpp` where I changed the building of the parameters when C++ code calls Python implemented regions.  The original logic did not allow nesting of parameter structures.   Calls to addRegion( ) would cause this to be called if the region is written in python.

I will appreciate any help I can get on this.

**Comment by dkeeney** [comment]
I gave up too soon...found the problem.
Of course it was someplace I was not expecting it to be...in the code that processed the Spec being read from the python code.  The changes in my code exposed a bug over in that code.
Anyway... I working on making the corrections....

**Comment by breznak** [comment]
Do you intend to keep options for both libraries, or we decide for one over another?

**Comment by dkeeney** [comment]
> Do you intend to keep options for both libraries, or we decide for one over another?

I don't know yet.
- yaml-cpp still has the crash problem.  
- libyaml does not have the crash problem.
- yaml-cpp is based on YAML 1.2
- libyaml is based on YAML 1.1

At no time will we have both libraries compiled at the same time.
But I am still doing A - B tests to see which one seems to work best for us.
Eventually I will probably decide on one and remove the interface for the other but for now I would like to keep the options open.

**Comment by dkeeney** [comment]
There are still some things I do not like about the Value class.
- its name, 'Value'.   I am not imaginative enough to come up with a better name.
- I use this 'assigned_' pointer which I do not like.  Perhaps there is a way to do this with shared_ptr.

At the moment I am working on PyBindRegion.cpp and trying to get it to work with the new Value class.  The GridCellLocationRegion.py uses parameters orientation and scale which are arrays of floats.  That was not handled correctly.   

But in general, the Spec limits the complexity of the parameter structure to flat mapped values or one dimensional arrays.  I would expect new Regions (or encoders) to need a much richer parameter structure.  But for now we are limited to this.

**Comment by dkeeney** [comment]
I want to tighten up the Value class before we merge.
I want to see if I can get rid of the assigned_ pointer. It might simplify a lot of code if the zombie object can be moved into the std::map without changing its address...so that things that point to it do not need to change their pointer after an assignment.

File: external/libYaml.cmake [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+#
+#            There is a problem in version 0.2.2; it does not build.
+#            Use the current master.
+#
+if(EXISTS   ${REPOSITORY_DIR}/build/ThirdParty/share/libyaml.zip)
+    set(URL ${REPOSITORY_DIR}/build/ThirdParty/share/libyaml.zip)
+elif(EXISTS ${REPOSITORY_DIR}/build/ThirdParty/share/libyaml.tar.gz)
+    set(URL ${REPOSITORY_DIR}/build/ThirdParty/share/libyaml.tar.gz)
+else()
+    set(URL "https://github.com/yaml/libyaml/archive/master.zip")
```
Comment (breznak) [inline external/libYaml.cmake]: please use a fixed release/tag/commit

File: src/htm/ntypes/Value.cpp [hunk 8]
```diff
@@ -15,271 +18,703 @@
...
+#include <stack>
+
 using namespace htm;
 
-Value::Value(std::shared_ptr<Scalar> &s) {
-  category_ = scalarCategory;
-  scalar_ = s;
-}
+#define ZOMBIE_MAP ((size_t)-1) // Means the key of zombie was a map key
+#define ZOMBIE_SEQ 0            // Means the key of zombie was a seq key
```
Comment (breznak) [inline src/htm/ntypes/Value.cpp]: what is this zombie terminology?

File: src/htm/ntypes/Value.hpp [hunk 9]
```diff
@@ -15,121 +17,386 @@
...
- * It is used internally in the conversion of YAML strings to C++ objects.
- * The API and implementation are geared towards clarify rather than
- * performance, since it is expected to be used only during network construction.
- */
-
+/***********************************************
+|  Value object - a container for data structures that have been parsed
+|  from a YAML or JSON string.  It can hold any structure that can be
+|  parsed from these but does not implement Alias or reference features.
+|  YAML is a superset of JSON format but we only use the JSON compatible features.
```
Comment (breznak) [inline src/htm/ntypes/Value.hpp]: could we use a JSON parser then? (not that we need, if libyaml works good)

File: external/libYaml.cmake (hunk 7 above)
Comment (dkeeney) [reply external/libYaml.cmake]: > please use a fixed release/tag/commit

The current release is 0..2.2 but it did not build.
so as mentioned in the comments, we have to use the master repository until they release again.

File: src/htm/ntypes/Value.cpp (hunk 8 above)
Comment (dkeeney) [reply src/htm/ntypes/Value.cpp]: > what is this zombie terminology?

I tried to describe it where I define operator[ ].  I will expand on that description.
Consider some code that looks like this:
```
   Value vm;
   vm.parse("{ hello: something}");
   Value& node = vm["hello"];
   node = vm["newNode"];
   node = "new value to be inserted into the tree";
```
When we do the lookup with the vm["hello"] it exists so we get a reference to a node in the tree that matches the key.

However, if the lookup did not find anything (as is the case with vm["newNode"]) we don't have anything to return a reference to.  What STL does is create an entry in the tree and return that but I would rather not change the tree until there is a value being assigned to it.  So what I do is create a zombie node, assign the key value to it,  and return a reference to it. This is just a Value node of type Empty which is not linked into the tree but holds the key and a pointer to the parent in case it is assigned to.

When the operator= makes the assignment, it then puts the value into the zombie, and then adds the zombie to the parent in the tree using the key, changing its type to Scalar. At that point it is no longer a zombie.

The problem is where does that node exist until a value can be assigned?  The caller has only a reference to it.  So what I do is put the zombie object on the parent node in a shared_ptr.  The down side is that there can be only one outstanding lookup (or zombie) per parent node that can be assigned to. Its scope lasts until the next lookup on that node.

A complication is that the caller could be holding a reference to that zombie node. When I add the zombie node to the internal map, it adds a copy so the address changes.  So that is what the assigned_ pointers are for.   They redirect all operations on a zombie to the real node after it has been assigned to.  By doing some shared_ptr magic I think I can get rid of that extra re-direction.

File: src/htm/ntypes/Value.hpp (hunk 9 above)
Comment (dkeeney) [reply src/htm/ntypes/Value.hpp]: > could we use a JSON parser 

Yes, we could.   But I would rather handle both JSON and Yaml.

File: external/libYaml.cmake (hunk 7 above)
Comment (breznak) [reply external/libYaml.cmake]: > we have to use the master repository until they release again. 

yes, but use specific commit that works, so the "version" is fixed

Comment (dkeeney) [reply external/libYaml.cmake]: How do I do that?  I think I can only have access to whatever is the latest in the archives.

**Comment by dkeeney** [comment]
I was able to get rid of the assigned_ feature so this is a little cleaner.  Not sure it is much simpler however.  

The zombies make it difficult to follow the code but I could not think of a better way to have a place to hold the keys and parent from a lookup until an assignment is made.  yaml-cpp and rapidyaml used zombies as well. libyaml did not have a tree.

While I was at it, I added some thread locks so I think this should be threadsafe.  I did not confirm that it is threadsafe...that would be a major test effort that is probably not worth it at this point.

**Comment by dkeeney** [comment]
The downloader for yaml-cpp is still in the package although not used.  I would like to keep that for a little while to make sure that libyaml continues to work ok for us.

With the addition of this package, the end users of the library should not see anything different.  However it does a few things for us.
- cleans up some rather ugly code.  
- encapsulates the yaml parser so that in the future we have the option of switching to a different yaml parser with minimum effort if we find there is a problem.
- Yaml (and JSON) parsing no longer uses the Spec so it can be used for things other than Region parameters.  The Value structure is a universal container for parsed Yaml.
- Moves the processing of parameters with the region Spec closer to the Region implementation so that the Region can have more flexibility in the layout of the parameters. It can now do nested sequences and maps.

In general it was much more work that should have been devoted to a low priority bug.  :-)

**Comment by dkeeney** [comment]
Oh, another idea as to how to simplify some more...
...the parent_ field points to the parent's core not the parent Value object.  
Let me see if that helps.

**Comment by dkeeney** [comment]
Yes, using the parent's core as the parent_ pointer on each node helped somewhat.
